File: backend/dormammu/state/execution_projection.py
from __future__ import annotations

from typing import Any, Mapping

from dormammu.results import RunResult, StageResult, latest_stage_results


def mutable_execution_block(state: Mapping[str, Any]) -> dict[str, Any]:
    block = state.get("execution")
    execution: dict[str, Any] = dict(block) if isinstance(block, Mapping) else {}
    stage_results = execution.get("stage_results")
    if isinstance(stage_results, Mapping):
        execution["stage_results"] = {
            str(key): dict(value)
            for key, value in stage_results.items()
            if isinstance(key, str) and isinstance(value, Mapping)
        }
    else:
        execution["stage_results"] = {}
    return execution


def project_stage_result(
    state: dict[str, Any],
    *,
    stage: StageResult,
    run_id: str | None,
    timestamp: str,
) -> None:
    stage_payload = stage.to_dict(include_output=False)
    stage_payload["recorded_at"] = timestamp
    effective_run_id = (
        run_id
        or stage.metadata.get("run_id")
        or stage.metadata.get("pipeline_run_id")
        or stage.metadata.get("latest_run_id")
    )
    state["updated_at"] = timestamp
    execution = mutable_execution_block(state)
    if effective_run_id is not None:
        execution["latest_run_id"] = effective_run_id
        execution["latest_execution_id"] = effective_run_id
    stage_results = execution.setdefault("stage_results", {})
    stage_results[stage.key] = stage_payload
    execution["latest_stage_result"] = stage_payload
    state["execution"] = execution


def project_run_result(
    state: dict[str, Any],
    *,
    result: RunResult,
    run_id: str | None,
    timestamp: str,
) -> None:
    effective_run_id = run_id or result.latest_run_id
    latest_by_stage = {
        stage.key: stage.to_dict(include_output=False)
        for stage in latest_stage_results(result.stage_results)
    }
    latest_run = result.to_dict(include_output=False)
    latest_run["run_id"] = effective_run_id
    latest_run["execution_run_id"] = effective_run_id
    latest_run["recorded_at"] = timestamp

    state["updated_at"] = timestamp
    execution = mutable_execution_block(state)
    execution["latest_run_id"] = effective_run_id
    execution["latest_execution_id"] = effective_run_id
    execution["current_run"] = None
    execution["latest_run"] = latest_run
    execution["stage_results"] = latest_by_stage
    execution["latest_stage_result"] = (
        latest_by_stage[next(reversed(latest_by_stage))]
        if latest_by_stage
        else None
    )
    state["execution"] = execution


def project_agent_run_fact(
    state: dict[str, Any],
    *,
    status: str,
    run_payload: Mapping[str, Any],
    timestamp: str,
) -> None:
    execution = mutable_execution_block(state)
    runtime_run = {
        "run_id": run_payload.get("run_id"),
        "status": status,
        "prompt_mode": run_payload.get("prompt_mode"),
        "workdir": run_payload.get("workdir"),
        "command": run_payload.get("command"),
        "exit_code": run_payload.get("exit_code"),
        "started_at": run_payload.get("started_at"),
        "completed_at": run_payload.get("completed_at"),
        "artifacts": run_payload.get("artifact_refs", []),
        "updated_at": timestamp,
    }
    execution["latest_run_id"] = run_payload.get("run_id")
    if status == "started":
        execution["current_run"] = runtime_run
    else:
        execution["current_run"] = None
        execution["latest_agent_run"] = runtime_run
    state["execution"] = execution


def project_lifecycle_execution_fact(
    state: dict[str, Any],
    *,
    event_payload: Mapping[str, Any],
    timestamp: str,
) -> None:
    execution = mutable_execution_block(state)
    event_type = str(event_payload.get("event_type") or "")
    payload = event_payload.get("payload")
    if not isinstance(payload, Mapping):
        payload = {}
    run_id = event_payload.get("run_id")
    if run_id is not None:
        execution["latest_run_id"] = run_id
        execution["latest_execution_id"] = run_id

    if event_type == "run.requested":
        execution["current_run"] = {
            "run_id": run_id,
            "status": event_payload.get("status"),
            "source": payload.get("source"),
            "entrypoint": payload.get("entrypoint"),
            "trigger": payload.get("trigger"),
            "prompt_summary": payload.get("prompt_summary"),
            "requested_at": timestamp,
        }
    elif event_type == "run.started":
        current_run = execution.get("current_run")
        next_current = dict(current_run) if isinstance(current_run, Mapping) else {}
        next_current.update(
            {
                "run_id": run_id,
                "status": event_payload.get("status"),
                "source": payload.get("source"),
                "entrypoint": payload.get("entrypoint"),
                "trigger": payload.get("trigger"),
                "started_at": timestamp,
            }
        )
        execution["current_run"] = next_current
    elif event_type == "run.finished":
        current_run = execution.get("current_run")
        latest_run = dict(current_run) if isinstance(current_run, Mapping) else {}
        latest_run.update(
            {
                "run_id": run_id,
                "status": event_payload.get("status"),
                "source": payload.get("source"),
                "entrypoint": payload.get("entrypoint"),
                "attempts_completed": payload.get("attempts_completed"),
                "retries_used": payload.get("retries_used"),
                "supervisor_verdict": payload.get("supervisor_verdict"),
                "outcome": payload.get("outcome"),
                "error": payload.get("error"),
                "artifacts": event_payload.get("artifact_refs", []),
                "completed_at": timestamp,
            }
        )
        execution["current_run"] = None
        execution["latest_run"] = latest_run
    elif event_type in {"stage.completed", "stage.failed"}:
        metadata = event_payload.get("metadata")
        stage_result = metadata.get("stage_result") if isinstance(metadata, Mapping) else None
        if not isinstance(stage_result, Mapping):
            stage_result = {
                "role": event_payload.get("role"),
                "stage_name": event_payload.get("stage") or event_payload.get("role"),
                "status": event_payload.get("status"),
                "verdict": payload.get("verdict"),
                "summary": payload.get("reason"),
                "artifacts": event_payload.get("artifact_refs", []),
            }
        stage_results = execution.setdefault("stage_results", {})
        stage_name = str(stage_result.get("stage_name") or stage_result.get("role") or "stage")
        stage_results[stage_name] = dict(stage_result)
        execution["latest_stage_result"] = dict(stage_result)
    elif event_type == "evaluator.checkpoint_decision":
        execution["latest_checkpoint"] = {
            "run_id": run_id,
            "stage": event_payload.get("stage"),
            "status": event_payload.get("status"),
            "checkpoint_kind": payload.get("checkpoint_kind"),
            "decision": payload.get("decision"),
            "rationale": payload.get("rationale"),
            "artifacts": event_payload.get("artifact_refs", []),
            "updated_at": timestamp,
        }
    elif event_type == "artifact.persisted":
        execution["latest_artifact"] = {
            "run_id": run_id,
            "stage": event_payload.get("stage"),
            "role": event_payload.get("role"),
            "artifact_kind": payload.get("artifact_kind"),
            "operation": payload.get("operation"),
            "summary": payload.get("summary"),
            "artifacts": event_payload.get("artifact_refs", []),
            "updated_at": timestamp,
        }

    state["execution"] = execution

File: runtime/src/state/executionProjection.ts
import {
  latestStageResults,
  runResultToDict,
  stageResultKey,
  stageResultToDict,
  type RunResult,
  type StageResult
} from "../results.js";

export type JsonRecord = Record<string, unknown>;

function isRecord(value: unknown): value is JsonRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function firstTruthyString(...values: readonly unknown[]): string | null {
  for (const value of values) {
    if (value) {
      return String(value);
    }
  }
  return null;
}

export function mutableExecutionBlock(state: Readonly<JsonRecord>): JsonRecord {
  const block = state.execution;
  const execution: JsonRecord = isRecord(block) ? { ...block } : {};
  const stageResults = execution.stage_results;
  if (isRecord(stageResults)) {
    const normalized: JsonRecord = {};
    for (const [key, value] of Object.entries(stageResults)) {
      if (isRecord(value)) {
        normalized[key] = { ...value };
      }
    }
    execution.stage_results = normalized;
  } else {
    execution.stage_results = {};
  }
  return execution;
}

export function projectStageResult(
  state: JsonRecord,
  options: { stage: StageResult; runId?: string | null; timestamp: string }
): void {
  const stagePayload = stageResultToDict(options.stage, { includeOutput: false });
  stagePayload.recorded_at = options.timestamp;
  const metadata = options.stage.metadata ?? {};
  const effectiveRunId = firstTruthyString(
    options.runId,
    metadata.run_id,
    metadata.pipeline_run_id,
    metadata.latest_run_id
  );

  state.updated_at = options.timestamp;
  const execution = mutableExecutionBlock(state);
  if (effectiveRunId !== null) {
    execution.latest_run_id = effectiveRunId;
    execution.latest_execution_id = effectiveRunId;
  }
  const stageResults = execution.stage_results;
  if (isRecord(stageResults)) {
    stageResults[stageResultKey(options.stage)] = stagePayload;
  }
  execution.latest_stage_result = stagePayload;
  state.execution = execution;
}

export function projectRunResult(
  state: JsonRecord,
  options: { result: RunResult; runId?: string | null; timestamp: string }
): void {
  const effectiveRunId = firstTruthyString(
    options.runId,
    options.result.latestRunId,
    options.result.latest_run_id
  );
  const stageResults = options.result.stageResults ?? options.result.stage_results ?? [];
  const latestByStage: JsonRecord = {};
  for (const stage of latestStageResults(stageResults)) {
    latestByStage[stageResultKey(stage)] = stageResultToDict(stage, { includeOutput: false });
  }

  const latestRun = runResultToDict(options.result, { includeOutput: false });
  latestRun.run_id = effectiveRunId;
  latestRun.execution_run_id = effectiveRunId;
  latestRun.recorded_at = options.timestamp;

  state.updated_at = options.timestamp;
  const execution = mutableExecutionBlock(state);
  execution.latest_run_id = effectiveRunId;
  execution.latest_execution_id = effectiveRunId;
  execution.current_run = null;
  execution.latest_run = latestRun;
  execution.stage_results = latestByStage;
  const latestStagePayloads = Object.values(latestByStage);
  execution.latest_stage_result = latestStagePayloads.length
    ? latestStagePayloads[latestStagePayloads.length - 1]
    : null;
  state.execution = execution;
}

export function projectAgentRunFact(
  state: JsonRecord,
  options: { status: string; runPayload: Readonly<JsonRecord>; timestamp: string }
): void {
  const execution = mutableExecutionBlock(state);
  const runtimeRun = {
    run_id: options.runPayload.run_id,
    status: options.status,
    prompt_mode: options.runPayload.prompt_mode,
    workdir: options.runPayload.workdir,
    command: options.runPayload.command,
    exit_code: options.runPayload.exit_code,
    started_at: options.runPayload.started_at,
    completed_at: options.runPayload.completed_at,
    artifacts: options.runPayload.artifact_refs ?? [],
    updated_at: options.timestamp
  };
  execution.latest_run_id = options.runPayload.run_id;
  if (options.status === "started") {
    execution.current_run = runtimeRun;
  } else {
    execution.current_run = null;
    execution.latest_agent_run = runtimeRun;
  }
  state.execution = execution;
}

export function projectLifecycleExecutionFact(
  state: JsonRecord,
  options: { eventPayload: Readonly<JsonRecord>; timestamp: string }
): void {
  const execution = mutableExecutionBlock(state);
  const eventType = String(options.eventPayload.event_type ?? "");
  const payload = isRecord(options.eventPayload.payload) ? options.eventPayload.payload : {};
  const runId = options.eventPayload.run_id;
  if (runId != null) {
    execution.latest_run_id = runId;
    execution.latest_execution_id = runId;
  }

  if (eventType === "run.requested") {
    execution.current_run = {
      run_id: runId,
      status: options.eventPayload.status,
      source: payload.source,
      entrypoint: payload.entrypoint,
      trigger: payload.trigger,
      prompt_summary: payload.prompt_summary,
      requested_at: options.timestamp
    };
  } else if (eventType === "run.started") {
    const currentRun = isRecord(execution.current_run) ? execution.current_run : {};
    execution.current_run = {
      ...currentRun,
      run_id: runId,
      status: options.eventPayload.status,
      source: payload.source,
      entrypoint: payload.entrypoint,
      trigger: payload.trigger,
      started_at: options.timestamp
    };
  } else if (eventType === "run.finished") {
    const currentRun = isRecord(execution.current_run) ? execution.current_run : {};
    execution.current_run = null;
    execution.latest_run = {
      ...currentRun,
      run_id: runId,
      status: options.eventPayload.status,
      source: payload.source,
      entrypoint: payload.entrypoint,
      attempts_completed: payload.attempts_completed,
      retries_used: payload.retries_used,
      supervisor_verdict: payload.supervisor_verdict,
      outcome: payload.outcome,
      error: payload.error,
      artifacts: options.eventPayload.artifact_refs ?? [],
      completed_at: options.timestamp
    };
  } else if (eventType === "stage.completed" || eventType === "stage.failed") {
    const metadata = isRecord(options.eventPayload.metadata) ? options.eventPayload.metadata : {};
    const metadataStageResult = metadata.stage_result;
    const stageResult = isRecord(metadataStageResult)
      ? { ...metadataStageResult }
      : {
          role: options.eventPayload.role,
          stage_name: options.eventPayload.stage ?? options.eventPayload.role,
          status: options.eventPayload.status,
          verdict: payload.verdict,
          summary: payload.reason,
          artifacts: options.eventPayload.artifact_refs ?? []
        };
    const stageResults = isRecord(execution.stage_results) ? execution.stage_results : {};
    const stageName = String(stageResult.stage_name ?? stageResult.role ?? "stage");
    stageResults[stageName] = { ...stageResult };
    execution.stage_results = stageResults;
    execution.latest_stage_result = { ...stageResult };
  } else if (eventType === "evaluator.checkpoint_decision") {
    execution.latest_checkpoint = {
      run_id: runId,
      stage: options.eventPayload.stage,
      status: options.eventPayload.status,
      checkpoint_kind: payload.checkpoint_kind,
      decision: payload.decision,
      rationale: payload.rationale,
      artifacts: options.eventPayload.artifact_refs ?? [],
      updated_at: options.timestamp
    };
  } else if (eventType === "artifact.persisted") {
    execution.latest_artifact = {
      run_id: runId,
      stage: options.eventPayload.stage,
      role: options.eventPayload.role,
      artifact_kind: payload.artifact_kind,
      operation: payload.operation,
      summary: payload.summary,
      artifacts: options.eventPayload.artifact_refs ?? [],
      updated_at: options.timestamp
    };
  }

  state.execution = execution;
}
